File: csharp/BookInvoicing/Purchase/Invoice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BookInvoicing.Domain.Country;
using BookInvoicing.Finance;

namespace BookInvoicing.Purchase
{
    public sealed class Invoice
    {
        public int Id { get; }
        public string ClientName { get; }
        public Country Country { get; }
        public List<PurchasedBook> PurchasedBooks { get; }


        public Invoice(string clientName, Country country)
            : this(IdGenerator.NextId(), clientName, country)
        {
        }

        public Invoice(int id, string clientName, Country country)
        {
            Id = id;
            ClientName = clientName;
            Country = country;
            PurchasedBooks = new List<PurchasedBook>();
        }

        public void AddPurchasedBooks(List<PurchasedBook> books)
        {
            PurchasedBooks.AddRange(books);
        }

        public void AddPurchasedBook(PurchasedBook book)
        {
            PurchasedBooks.Add(book);
        }

        public double ComputeTotalAmount()
        {
            var totalAmount = 0.0;
            totalAmount = PurchasedBooks.Sum(book => book.TotalPrice * TaxRule.GetApplicableRate(Country, book.Book));
            return totalAmount;
        }

        public override string ToString() => $"Invoice [ {nameof(Id)}: '{Id}'" +
                                             $", {nameof(ClientName)}: '{ClientName}'" +
                                             $", {nameof(Country)}: '{Country}'" +
                                             $", {nameof(PurchasedBooks)}: '{PurchasedBooks}' ]";

        private bool Equals(Invoice other) => Id == other.Id
                                              && ClientName == other.ClientName
                                              && Equals(Country, other.Country)
                                              && Equals(PurchasedBooks, other.PurchasedBooks);

        public void AddPurchasedBooks(object books)
        {
            throw new NotImplementedException();
        }

        public override bool Equals(object obj) => ReferenceEquals(this, obj)
                                                   || obj is Invoice other && Equals(other);

        public override int GetHashCode() => HashCode.Combine(Id, ClientName, Country, PurchasedBooks);
    }
}

File: go/purchase/invoice.go
package purchase

import (
	. "book_invoicing/domain/country"
	. "book_invoicing/finance"
	. "book_invoicing/service"
)

type Invoice struct {
	ID             int64
	ClientName     string
	Country        Country
	PurchasedBooks []PurchasedBook
}

func NewInvoice(clientName string, country Country, id ...int64) *Invoice {
	invoiceID := NextID()
	if len(id) > 0 {
		invoiceID = id[0]
	}
	return &Invoice{
		ID:             invoiceID,
		ClientName:     clientName,
		Country:        country,
		PurchasedBooks: []PurchasedBook{},
	}
}

func (i *Invoice) AddPurchasedBook(purchasedBook PurchasedBook) {
	i.PurchasedBooks = append(i.PurchasedBooks, purchasedBook)
}

func (i *Invoice) AddPurchasedBooks(purchasedBooks []PurchasedBook) {
	for _, pb := range purchasedBooks {
		i.AddPurchasedBook(pb)
	}
}

func (i *Invoice) ComputeTotalAmount() float64 {
	var sum float64
	for _, pb := range i.PurchasedBooks {
		totalPrice := pb.TotalPrice() * GetApplicableRate(i.Country, pb.Book)
		sum += totalPrice
	}
	return sum
}
